src/test/futex/test_futex.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+#include <errno.h>
+#include <linux/futex.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/syscall.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+// Note: this test roughly follows the example at the end of `man 2 futex`
+
+#define PTR_TO_INT(p) ((int)(long)(p))
+#define INT_TO_PTR(i) ((void*)(long)(i))
+#define NUM_LOOPS 100
+
+// The futex word used to synchronize threads
+volatile int futex_word1 = 0; // initial state is unavailable
+volatile int futex_word2 = 1; // initial state is available
+
+// Acquire: wait for the futex pointed to by `word` to become 1, then set to 0
+static int _futex_wait(volatile int* word) {
+    while (1) {
+        // Args are: ptr, expected old val, desired new val
+        bool is_available = 1 == __sync_val_compare_and_swap(word, 1, 0);
+
+        if (is_available) {
+            break;
+        } else {
+            int res = syscall(SYS_futex, word, FUTEX_WAIT, 0, NULL, NULL, 0);
+            if (res == -1 && errno != EAGAIN) {
```

Comment:
If I'm right about the bug in the `futex` handler, then after woken we'll have EAGAIN and continue to loop; presumably forever. Was thinking maybe this is the "shadow exits without error even if we had to kill the plugin violently" problem, but since you pthread_join the child threads shouldn't that hang and cause test to timeout? Maybe pthread_join is returning before it should due to some other bug or missing functionality...?

Possible alternative to relying on pthread_join is to busy wait on some volatile "finished" flag as in the clone and (detached) pthread test. Seems like that'd be reasonable to do here since pthread_join itself depends on futex, so could cause confusing behavior here if futex isn't working correctly.

Reply:
Hmm... I'm not exactly sure what was going on there. I thought the atomics ops were preventing any infinite loops. In any case, I stopped using `pthread_join` in favor of the busy loop (as is done in the clone test) as suggested.